commit 6f395451e9c61d50ce868de7493af4ad45206ac4 (refs/changes/63/3163/2)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-01-05T14:54:11-08:00 (1 year, 9 months ago)
    
    CNAPI-732 document that ZFS API functions are deprecated
    Reviewed by: Orlando Vazquez <orlando@joyent.com>
    Approved by: Orlando Vazquez <orlando@joyent.com>

diff --git a/docs/index.md b/docs/index.md
index 0bffb63..00ac9aa 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -1,6 +1,6 @@
 ---
 title: CNAPI (Compute Node API) Design
-apisections: Allocation API, Boot Parameters API, Compute Node Agent Tasks API, Miscellaneous API, Remote Execution API, Server API, Virtual Machine API, Virtual Machine Images API, Virtual Machine Snapshots API, Waitlist API, ZFS API
+apisections: Allocation API, Boot Parameters API, Compute Node Agent Tasks API, Miscellaneous API, Remote Execution API, Server API, Virtual Machine API, Virtual Machine Images API, Virtual Machine Snapshots API, Waitlist API, ZFS API (deprecated)
 markdown2extras: tables, code-friendly
 ---
 <!--
@@ -10,7 +10,7 @@ markdown2extras: tables, code-friendly
 -->
 
 <!--
-    Copyright (c) 2016, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <!--
@@ -1881,9 +1881,12 @@ None.
 
 
 
-# ZFS API
+# ZFS API (deprecated)
 
-## DatasetsList (GET GET /servers/:server_uuid/datasets)
+## DatasetsList (GET /servers/:server_uuid/datasets)
+
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
 
 List ZFS datasets on a server.
 
@@ -1901,6 +1904,9 @@ None.
 
 ## DatasetCreate (POST /servers/:server_uuid/datasets)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Create a ZFS dataset on a server.
 
 ### Inputs
@@ -1917,6 +1923,9 @@ None.
 
 ## SnapshotCreate (POST /servers/:server_uuid/datasets/:dataset/snapshot)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Create a ZFS snapshot of a dataset on a server.
 
 ### Inputs
@@ -1935,6 +1944,9 @@ Create a ZFS snapshot of a dataset on a server.
 
 ## SnapshotRollback (POST /servers/:server_uuid/datasets/:dataset/rollback)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Revert a ZFS dataset to back to a previous state captured by a snapshot.
 
 ### Inputs
@@ -1953,6 +1965,9 @@ Revert a ZFS dataset to back to a previous state captured by a snapshot.
 
 ## SnapshotList (GET /servers/:server_uuid/datasets/:dataset/snapshots)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 List all snapshots on a dataset
 
 ### Inputs
@@ -1969,6 +1984,9 @@ None.
 
 ## DatasetPropertiesGetAll (GET /servers/:server_uuid/dataset-properties)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Get ZFS properties across all datasets on a server.
 
 ### Inputs
@@ -1989,6 +2007,9 @@ Get ZFS properties across all datasets on a server.
 
 ## DatasetPropertiesGet (GET /servers/:server_uuid/datasets/:dataset/properties)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Get ZFS properties for a dataset.  The specific properties to return can be
 filtered with ?prop1=foo&prop2=bar, etc.
 
@@ -2010,6 +2031,9 @@ filtered with ?prop1=foo&prop2=bar, etc.
 
 ## DatasetPropertiesSet (POST /servers/:server_uuid/datasets/:dataset/properties)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Set one or more properties for a ZFS dataset.
 
 ### Inputs
@@ -2028,6 +2052,9 @@ Set one or more properties for a ZFS dataset.
 
 ## DatasetDestroy (DELETE /servers/:server_uuid/datasets/:dataset)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 Destroy a ZFS dataset on a server.
 
 ### Inputs
@@ -2044,6 +2071,9 @@ None.
 
 ## ZpoolList (GET /servers/:server_uuid/zpools)
 
+*IMPORTANT: This endpoint is deprecated and will be removed in a future
+release. Do not use.*
+
 List the ZFS pools on a server.
 
 ### Inputs
diff --git a/docs/index/index.md.ejs b/docs/index/index.md.ejs
index aa39a09..bf4b8a4 100644
--- a/docs/index/index.md.ejs
+++ b/docs/index/index.md.ejs
@@ -10,7 +10,7 @@ markdown2extras: tables, code-friendly
 -->
 
 <!--
-    Copyright (c) 2016, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <!--
diff --git a/lib/endpoints/zfs.js b/lib/endpoints/zfs.js
index 69a0ac7..d301cb2 100644
--- a/lib/endpoints/zfs.js
+++ b/lib/endpoints/zfs.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -19,13 +19,22 @@ var ModelServer = require('../models/server');
 
 function ZFS() {}
 
+function logDeprecated(req) {
+    req.log.error({
+        req: req,
+        route: req.route.name
+    }, 'WARNING: deprecated endpoint called.');
+}
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * List ZFS datasets on a server.
  *
  * @name DatasetsList
- * @endpoint GET GET /servers/:server_uuid/datasets
- * @section ZFS API
+ * @endpoint GET /servers/:server_uuid/datasets
+ * @section ZFS API (deprecated)
  *
  * @example GET /servers/44454c4c-4800-1034-804a-b2c04f354d31/datasets
  *
@@ -39,6 +48,9 @@ ZFS.listDatasets = function handlerZfsListDatasets(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_list_datasets', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
@@ -52,11 +64,14 @@ ZFS.listDatasets = function handlerZfsListDatasets(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Create a ZFS dataset on a server.
  *
  * @name DatasetCreate
  * @endpoint POST /servers/:server_uuid/datasets
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @example POST /servers/44454c4c-4800-1034-804a-b2c04f354d31/datasets
  *         -d "datasets=zones/myfs"
@@ -71,6 +86,9 @@ ZFS.createDataset = function handlerZfsCreateDataset(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_create_dataset', opts, function (err) {
         if (err) {
             return (next(new restify.InternalError(
@@ -84,11 +102,14 @@ ZFS.createDataset = function handlerZfsCreateDataset(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Create a ZFS snapshot of a dataset on a server.
  *
  * @name SnapshotCreate
  * @endpoint POST /servers/:server_uuid/datasets/:dataset/snapshot
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @param {String} name The name of the snapshot to create
  *
@@ -108,6 +129,9 @@ ZFS.createSnapshot = function handlerZfsCreateSnapshot(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask(
         'zfs_snapshot_dataset',
         opts,
@@ -124,11 +148,14 @@ ZFS.createSnapshot = function handlerZfsCreateSnapshot(req, res, next) {
 };
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Revert a ZFS dataset to back to a previous state captured by a snapshot.
  *
  * @name SnapshotRollback
  * @endpoint POST /servers/:server_uuid/datasets/:dataset/rollback
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @param {String} name The name of the snapshot to be created
  *
@@ -149,6 +176,9 @@ ZFS.rollbackSnapshot = function handlerZfsRollbackSnapshot(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask(
         'zfs_rollback_dataset',
         opts,
@@ -166,11 +196,14 @@ ZFS.rollbackSnapshot = function handlerZfsRollbackSnapshot(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * List all snapshots on a dataset
  *
  * @name SnapshotList
  * @endpoint GET /servers/:server_uuid/datasets/:dataset/snapshots
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @example GET /servers/44454c4c-4800-1034-804a-b2c04f354d31/datasets
  *
@@ -185,6 +218,9 @@ ZFS.listSnapshots = function handlerZfsListSnapshots(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_list_snapshots', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
@@ -212,11 +248,14 @@ function findProps(params) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Get ZFS properties across all datasets on a server.
  *
  * @name DatasetPropertiesGetAll
  * @endpoint GET /servers/:server_uuid/dataset-properties
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @param {String} <prop1> Get the property given by the "prop1" value
  * @param {String} <prop2> Get the property given by the "prop2" value
@@ -240,6 +279,9 @@ ZFS.getAllProperties = function handlerZfsGetAllProperties(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_get_properties', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
@@ -256,12 +298,15 @@ ZFS.getAllProperties = function handlerZfsGetAllProperties(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Get ZFS properties for a dataset.  The specific properties to return can be
  * filtered with ?prop1=foo&prop2=bar, etc.
  *
  * @name DatasetPropertiesGet
  * @endpoint GET /servers/:server_uuid/datasets/:dataset/properties
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @param {String} <prop1> Get the property given by the "prop1" value
  * @param {String} <prop2> Get the property given by the "prop2" value
@@ -291,6 +336,9 @@ ZFS.getProperties = function handlerZfsGetProperties(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_get_properties', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
@@ -307,11 +355,14 @@ ZFS.getProperties = function handlerZfsGetProperties(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Set one or more properties for a ZFS dataset.
  *
  * @name DatasetPropertiesSet
  * @endpoint POST /servers/:server_uuid/datasets/:dataset/properties
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @param {Object} properties Object containing string property values
  *
@@ -334,6 +385,9 @@ ZFS.setProperties = function handlerZfsSetProperties(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_set_properties', opts, function (err) {
         if (err) {
             return (next(new restify.InternalError(
@@ -347,11 +401,14 @@ ZFS.setProperties = function handlerZfsSetProperties(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * Destroy a ZFS dataset on a server.
  *
  * @name DatasetDestroy
  * @endpoint DELETE /servers/:server_uuid/datasets/:dataset
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @example DELETE /servers/44454c4c-4800-1034-804a-b2c04f354d31
  *          /datasets/zones%2fmyfs
@@ -367,6 +424,8 @@ ZFS.destroyDataset = function handlerZfsDestroyDataset(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
 
     server.zfsTask('zfs_destroy_dataset', opts, function (err) {
         if (err) {
@@ -381,11 +440,14 @@ ZFS.destroyDataset = function handlerZfsDestroyDataset(req, res, next) {
 
 
 /**
+ * *IMPORTANT: This endpoint is deprecated and will be removed in a future
+ * release. Do not use.*
+ *
  * List the ZFS pools on a server.
  *
  * @name ZpoolList
  * @endpoint GET /servers/:server_uuid/zpools
- * @section ZFS API
+ * @section ZFS API (deprecated)
  *
  * @example GET /servers/44454c4c-4800-1034-804a-b2c04f354d31/zpools
  *
@@ -400,6 +462,9 @@ ZFS.listZpools = function handlerZfsListZpools(req, res, next) {
         req: req
     };
 
+    // This endpoint is deprecated, log that it was called.
+    logDeprecated(req);
+
     server.zfsTask('zfs_list_pools', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
